### ISSUE TITLE
Add "solo" button to resource versions

### ISSUE THREAD

**Issue by krishicks** [concourse/concourse#1288]
Sometimes I want to target a specific version of a resource when running a pipeline, but I don't want to have to re-set the pipeline, modifying the `get` of that resource to have a specific version. It's a temporary thing, and I don't want to block other users of the pipeline (which I also cannot do other than verbally), nor do I want to have to spin up a new pipeline just for this purpose; I want the history in the same pipeline.

Presently to do this I have to disable every other newer version of the resource, which becomes unwieldy when there are many versions of the resource.

Similar to how a mixer has buttons next to each channel for "mute" and "solo", it would be nice to be able to  solo a resource in addition to the existing "mute" feature.

**Comment by vito** [comment]
Heyo! I just opened https://github.com/concourse/concourse/issues/2386 as a prelude to the "spaces" epic, and one implementation detail touches on this subject. The main context is at the bottom but it may help to read the whole thing.

tl;dr: do you think we could just replace "resource pausing" with this feature? I think if we have the UX clear enough by e.g. showing the pinned version in some special way on the resource page and from the pipeline view, it'll actually be an improvement over the current "disable everything and then pause" workflow.

**Comment by krishicks** [comment]
@vito It does sound similar. I don't think I've ever paused the whole resource before; usually I pause a bunch of versions because I need to go back to an earlier one.

I think the solo feature would be better, though I wonder how it would impact people that do use the resource pausing feature to prevent new versions from being detected.

**Comment by krishicks** [comment]
It might be interesting also to provide a new interface in the build page.

A question that has come up from my team here is "what versions will be used if I push `+` on a build page?"

Presently the only way to figure this out is to click through each resource and see what the latest version is, and that may not be the actual version when you click the `+` because of the check that occurs when manually triggering a build.

This could be improved by showing what the currently-detected versions are for each resource on the build page. That would need to live-update, of course. Additionally, if you could push a button to change the displayed version into a drop-down it could give you the option to choose what versions to use. I am not sure how that would work, given a "version" is potentially a long string of text with resource-specific metadata. Maybe it'd be a potentially-huge (e.g. very wide) dropdown. I'm thinking of docker image digests here.

**Comment by krishicks** [comment]
Oh, and the version shown in the build page wouldn't be the latest version detected, but the latest version that could be used for a new build, meaning all the passed constraints apply.

**Comment by vito** [comment]
@krishicks That second proposal sounds more like https://github.com/concourse/concourse/issues/269. There's also some context in https://github.com/concourse/concourse/issues/413#issuecomment-233076618.

**Comment by xdom** [comment]
Does this solution resolve race condition when multiple users would "solo" (unmute) different versions?

**Comment by krishicks** [comment]
Solo isn't meant to unmute; it's meant to mute everything else. Any race conditions would still remain.

**Comment by vito** [comment]
I think we can satisfy https://github.com/concourse/docker-image-resource/pull/229, https://github.com/concourse/docker-image-resource/issues/194, and https://github.com/concourse/docker-image-resource/issues/187 by also allowing this to be configurable in the pipeline.

Then the button in the web UI would just correspond to setting that same config. We could even have it be preserved in `fly get-pipeline`. (I don't want to go down the rabbit hole of having the web UI be a full pipeline configurator, but this seems like pretty simple quality-of-life behavior.)

**Comment by vito** [comment]
This is the first time we'll have web UI functionality overlap with pipeline config, so it's a bit fun to reason about the precedence. I chatted with @clarafu and @mhuangpivotal and here's what we came up with:

* If I have a resource pinned in the *config*, I should *not* be able to pin other versions or unpin it in the web UI. The config takes precedence, and the web UI should reflect that.
* If I have a resource pinned in the *web UI*, and then pin it in the *config* (via `set-pipeline`), it should "clear out" whatever state made it pinned in the web UI, and make it pinned via the config (again, removing the ability to pin it in the web UI). Again, config takes precedence.
* If I have a resource pinned in the *web UI*, and run `set-pipeline` *without* it pinned, it should remain pinned via the web UI. This is to support "emergent" pinning and leave it unaffected by other, unrelated pipeline changes (say the PM/anchor pins in response to a broken dependency just before shipping, and then a dev runs `set-pipeline` as they're iterating on some other job).

**Comment by Lindsayauchin** [comment]
@vito @clarafu @mhuangpivotal 

Looking at possible UI solutions to expose this functionality in the UI I created a small invision prototype to show the flow of pinning/unpinning in the resource page and marking a resource as pinned in the pipeline.

**Prototype:** https://pivotal.invisionapp.com/share/8PNBM70CHN7#/311892305_02_Resource_Pin

## Resource page: 
![06_resource_pin_highlight_row_08](https://user-images.githubusercontent.com/27746614/43601120-7fdcb9aa-965a-11e8-9806-605edadd27a2.png)
^^ Pinned resources will have a highlight on the row and the ref will be reflected at the top of the page. Versions after that version will be dimmed down to show they are not active. 

![08_resource_pin_hover](https://user-images.githubusercontent.com/27746614/43601629-a34f4564-965b-11e8-8fc3-54249d20a5c7.png)

^^ If I have a resource pinned in the config, I should not be able to pin other versions or unpin it in the web UI. The config takes precedence, and the web UI should reflect that with the ref reflected at the top of the page and a tooltip to indicate the pin was made in the config. 

![03_resource_pin](https://user-images.githubusercontent.com/27746614/43601451-31ff60e2-965b-11e8-94e1-5776f89d8c0a.png)


## Pipeline: Notification with dropdown list and icon on the resource
![02_resource_pin](https://user-images.githubusercontent.com/27746614/43601134-8b31e596-965a-11e8-99b5-a4034759d4af.png)

![03_resource_pin_hover](https://user-images.githubusercontent.com/27746614/43601841-4397b4c0-965c-11e8-90e4-0aca4a732c17.png)
^^ onhover of the ref inthe drop down will highlight the pinned resources in the pipeline UI.

**Comment by krishicks** [comment]
Every aspect of the proposed design is great except for one thing: I think _all_ of the non-pinned versions should be grayed out, not just the ones newer than the current version.

**Comment by Lindsayauchin** [comment]
Trying another version with the actions grouped to the right of the resource version. 

![2_00_resource_pinned](https://user-images.githubusercontent.com/27746614/43800007-1add1070-9a5d-11e8-990e-b18def83a717.png)

**Comment by jama-pivotal** [comment]
The expected result of this story is to have an API and we can split out two stories: one for the web UI and one for Fly.

**Comment by marco-m** [comment]
I am confused here. Isn't this equivalent to resource pinning, to appear in 5.0 ?

**Comment by vito** [comment]
@marco-m yes, this is the resorce pinning feature (we just didn't call it 'soloing')
